x/incentives/keeper/distribute_test.go

```diff
@@ -868,3 +871,196 @@ func (s *KeeperTestSuite) TestGetPoolFromGaugeId() {
 		})
 	}
 }
+
+// TestFunctionalInternalExternalCLGauge is a functional test that covers more complex scenarios relating to distributing incentives through gauges
+// at the end of each epoch.
+//
+//
+// Testing strategy:
+// 1. Initialize variables.
+// 2. Setup CL pool and gauge (gauge automatically gets created at the end of CL pool creation).
+// 3. Create external no-lock gauges for CL pools
+// 4. Create Distribution records to incentivize internal CL no-lock gauges
+// 5. let epoch 1 pass
+//  	- Check that incentive record has been correctly created and gauge has been correctly updated.
+// 		- all perpetual gauges must finish distributing records
+// 6. Remove distribution records for internal incentives using HandleReplacePoolIncentivesProposal
+// 7. let epoch 2 pass
+// 		- check only external non-perpetual gauges with 2 epochs distributed
+// 		- check gauge has been correctly updated
+// 8. let epoch 3 pass
+// 		- nothing distributes as non-perpetual gauges with 2 epochs have ended and perpetual gauges have not been reloaded
+//		- nothing should change in terms of incentive records
+func (s *KeeperTestSuite) TestFunctionalInternalExternalCLGauge() {
+	// 1. Initialize variables
+	s.SetupTest()
+
+	requiredBalances := sdk.NewCoins(sdk.NewCoin("eth", sdk.NewInt(1_000_000_000)), sdk.NewCoin("usdc", sdk.NewInt(1_000_000_000)), sdk.NewCoin("uosmo", sdk.NewInt(1_000_000_000)))
+	internalGaugeCoins := sdk.NewCoins(sdk.NewCoin("stake", sdk.NewInt(750_000)))                                               // distributed full sum at epoch
+	externalGaugeCoins := sdk.NewCoins(sdk.NewCoin("eth", sdk.NewInt(1_000_000)), sdk.NewCoin("usdc", sdk.NewInt(1_000_000)))   // distributed full sum at epoch
+	halfOfExternalGaugeCoins := sdk.NewCoins(sdk.NewCoin("eth", sdk.NewInt(500_000)), sdk.NewCoin("usdc", sdk.NewInt(500_000))) // distributed at each epoch for non-perp gauge with numEpoch = 2
+
+	internalGaugeDecCoins := osmoutils.ConvertCoinsToDecCoins(internalGaugeCoins)
+	externalGaugeDecCoins := osmoutils.ConvertCoinsToDecCoins(externalGaugeCoins)
+	halfOfExternalGaugeDecCoins := osmoutils.ConvertCoinsToDecCoins(halfOfExternalGaugeCoins)
+
+	s.FundAcc(s.TestAccs[0], requiredBalances)
+	s.FundAcc(s.TestAccs[1], requiredBalances)
+	s.FundAcc(s.TestAccs[2], requiredBalances)
```

Comment:
It is not clear to me why we need to initialize all 3 accounts with the same balances. Could we make this stricter by only initializing the balances needed exactly for each account?

This is so that if something consumes more funds than expected, we would get an "insufficient balance" error

Reply:
On this point, I think some test helpers use one of the 3 accounts directly inside its function definition. I think we should be providing the account directly to the helper so that it is clear which one is being used and for what.

To be specific, I would change `ChangeNoLockExternalGauges` to create one gauge and take an account address as an input:
```go
func (s *KeeperTestSuite) CreateNoLockExternalGauge(gaugeCreator sdk.AccAddress, clPoolId,uint64, externalGaugeCoins sdk.Coins) (, uint64) {

	// Create 1 external no-lock gauge perpetual over 1 epochs MsgCreateGauge
	clPoolExternalGaugeId, err := s.App.IncentivesKeeper.CreateGauge(s.Ctx, true, gaugeCreator, externalGaugeCoins,
		lockuptypes.QueryCondition{
			LockQueryType: lockuptypes.NoLock,
		},
		s.Ctx.BlockTime(),
		1,
		clPoolId,
	)
	s.Require().NoError(err)

	return clPoolExternalGaugeId
}
```

The caller now needs to provide the account address to this function, making it clear how one or the other account is used

Reply:
aah i see yea that definitely cleaner